File: go/surface_test.go
// Live surface-coverage test (Go): cookies, navigation history, windows, W3C
// actions, screenshot, and execute_script return shapes against a real headless
// Chrome served by a local httptest server (so cookies/navigation have a real
// http:// origin). Skips if chromedriver is absent.
package selenium

import (
	"encoding/base64"
	"net/http"
	"net/http/httptest"
	"os/exec"
	"strconv"
	"testing"
	"time"
)

func TestLiveSurface(t *testing.T) {
	driverBin, err := exec.LookPath("chromedriver")
	if err != nil {
		t.Skip("chromedriver not on PATH")
	}

	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "text/html; charset=utf-8")
		if r.URL.Path == "/two" {
			_, _ = w.Write([]byte(`<!doctype html><title>Page Two</title><h1 id="hdr">Two</h1>`))
			return
		}
		_, _ = w.Write([]byte(`<!doctype html><title>Page One</title>` +
			`<h1 id="hdr">One</h1><a id="go" href="/two">to two</a>` +
			`<button id="btn" onclick="document.getElementById('hdr').textContent='clicked'">b</button>`))
	}))
	defer srv.Close()

	port := freePort(t)
	cmd := exec.Command(driverBin, "--port="+strconv.Itoa(port))
	if err := cmd.Start(); err != nil {
		t.Skipf("could not start chromedriver: %v", err)
	}
	defer func() { _ = cmd.Process.Kill(); _, _ = cmd.Process.Wait() }()
	if !waitUp(port, 10*time.Second) {
		t.Skip("chromedriver did not come up")
	}

	d, err := NewChrome("http://127.0.0.1:"+strconv.Itoa(port), Headless())
	if err != nil {
		t.Fatalf("NewChrome: %v", err)
	}
	defer d.Quit()

	if err := d.Get(srv.URL + "/one"); err != nil {
		t.Fatalf("Get: %v", err)
	}
	if title, _ := d.Title(); title != "Page One" {
		t.Fatalf("title = %q", title)
	}

	// navigation history
	go2, _ := d.FindElement(By.Id("go"))
	go2.Click()
	if title, _ := d.Title(); title != "Page Two" {
		t.Fatalf("after click title = %q", title)
	}
	d.Back()
	if title, _ := d.Title(); title != "Page One" {
		t.Fatalf("after back title = %q", title)
	}
	d.Forward()
	if title, _ := d.Title(); title != "Page Two" {
		t.Fatalf("after forward title = %q", title)
	}
	d.Back()

	// cookies
	d.DeleteAllCookies()
	if err := d.AddCookie(map[string]interface{}{"name": "flavor", "value": "mint"}); err != nil {
		t.Fatalf("AddCookie: %v", err)
	}
	c, err := d.Cookie("flavor")
	if err != nil {
		t.Fatalf("Cookie: %v", err)
	}
	if c.Value != "mint" {
		t.Fatalf("cookie value = %v", c.Value)
	}
	if err := d.DeleteCookie("flavor"); err != nil {
		t.Fatalf("DeleteCookie: %v", err)
	}

	// windows
	handles, _ := d.WindowHandles()
	if len(handles) < 1 {
		t.Fatalf("no window handles")
	}
	if _, err := d.SetWindowRect(map[string]interface{}{"width": 900, "height": 650}); err != nil {
		t.Fatalf("SetWindowRect: %v", err)
	}
	rect, _ := d.WindowRect()
	if rect.Width != 900 {
		t.Fatalf("window rect = %v", rect)
	}

	// execute_script shapes
	if v, _ := d.ExecuteScript("return 6*7;"); v.(float64) != 42 {
		t.Fatalf("script scalar = %v", v)
	}
	if v, _ := d.ExecuteScript("return arguments[0]+arguments[1];", 40, 2); v.(float64) != 42 {
		t.Fatalf("script args = %v", v)
	}

	// W3C actions: pointer click on the button.
	btn, _ := d.FindElement(By.Id("btn"))
	rct, _ := btn.Rect()
	cx := int(rct.X + rct.Width/2)
	cy := int(rct.Y + rct.Height/2)
	if err := d.PerformActions([]interface{}{map[string]interface{}{
		"type": "pointer", "id": "mouse", "parameters": map[string]interface{}{"pointerType": "mouse"},
		"actions": []interface{}{
			map[string]interface{}{"type": "pointerMove", "duration": 0, "x": cx, "y": cy},
			map[string]interface{}{"type": "pointerDown", "button": 0},
			map[string]interface{}{"type": "pointerUp", "button": 0},
		},
	}}); err != nil {
		t.Fatalf("PerformActions: %v", err)
	}
	hdr, _ := d.FindElement(By.Id("hdr"))
	if txt, _ := hdr.Text(); txt != "clicked" {
		t.Fatalf("actions click did not fire, hdr=%q", txt)
	}
	d.ClearActions()

	// screenshot -> PNG
	shot, _ := d.ScreenshotBase64()
	raw, err := base64.StdEncoding.DecodeString(shot)
	if err != nil || len(raw) < 8 || string(raw[1:4]) != "PNG" {
		t.Fatalf("screenshot not a PNG (%d bytes)", len(raw))
	}

	t.Logf("live surface: cookies/nav/windows/actions/screenshot/script all green")
}

// TestLiveFullFeatureSurface drives the full-feature additions against real
// headless Chrome: active element, new/close window, frame switching, element
// CSS value, form submit, print-to-PDF, Exists, and alert-present probing.
func TestLiveFullFeatureSurface(t *testing.T) {
	driverBin, err := exec.LookPath("chromedriver")
	if err != nil {
		t.Skip("chromedriver not on PATH")
	}
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "text/html; charset=utf-8")
		if r.URL.Path == "/framed" {
			_, _ = w.Write([]byte(`<!doctype html><title>Framed</title><p id="inner" style="color: rgb(0, 128, 0)">inside</p>`))
			return
		}
		_, _ = w.Write([]byte(`<!doctype html><title>Main</title>` +
			`<input id="fld" autofocus style="color: rgb(255, 0, 0)">` +
			`<form id="f" action="/framed"><input id="q" name="q"><button id="sub">go</button></form>` +
			`<iframe id="fr" src="/framed"></iframe>`))
	}))
	defer srv.Close()

	port := freePort(t)
	cmd := exec.Command(driverBin, "--port="+strconv.Itoa(port))
	if err := cmd.Start(); err != nil {
		t.Skipf("could not start chromedriver: %v", err)
	}
	defer func() { _ = cmd.Process.Kill(); _, _ = cmd.Process.Wait() }()
	if !waitUp(port, 10*time.Second) {
		t.Skip("chromedriver did not come up")
	}

	d, err := NewHeadlessChrome("http://127.0.0.1:" + strconv.Itoa(port))
	if err != nil {
		t.Fatalf("NewHeadlessChrome: %v", err)
	}
	defer d.Quit()
	if err := d.Get(srv.URL + "/main"); err != nil {
		t.Fatalf("Get: %v", err)
	}

	// Exists: present vs absent, no implicit wait.
	if ok, err := d.Exists(By.Id("fld")); err != nil || !ok {
		t.Fatalf("Exists(fld) = %v, %v; want true", ok, err)
	}
	if ok, err := d.Exists(By.Id("nope")); err != nil || ok {
		t.Fatalf("Exists(nope) = %v, %v; want false", ok, err)
	}

	// ActiveElement: the autofocus input.
	ae, err := d.ActiveElement()
	if err != nil {
		t.Fatalf("ActiveElement: %v", err)
	}
	if id, _, _ := ae.GetAttribute("id"); id != "fld" {
		t.Fatalf("active element id = %q; want fld", id)
	}

	// Element CSS value: computed display of the input is non-empty.
	fld, _ := d.FindElement(By.Id("fld"))
	if disp, err := fld.CssValue("display"); err != nil || disp == "" {
		t.Fatalf("CssValue(display) = %q, %v; want a non-empty computed value", disp, err)
	}

	// AlertPresent: none open.
	if present, err := d.AlertPresent(); err != nil || present {
		t.Fatalf("AlertPresent = %v, %v; want false", present, err)
	}

	// Frame switching: into the iframe, read a scoped element, back out.
	fr, _ := d.FindElement(By.Id("fr"))
	if err := d.SwitchToFrame(FrameElement(fr)); err != nil {
		t.Fatalf("SwitchToFrame: %v", err)
	}
	inner, err := d.FindElement(By.Id("inner"))
	if err != nil {
		t.Fatalf("find inside frame: %v", err)
	}
	if txt, _ := inner.Text(); txt != "inside" {
		t.Fatalf("inner text = %q; want inside", txt)
	}
	// CssValue exact match on a rendered styled element (via the alias too).
	if col, err := inner.ValueOfCssProperty("color"); err != nil || col != "rgba(0, 128, 0, 1)" {
		t.Fatalf("ValueOfCssProperty(color) = %q, %v; want rgba(0, 128, 0, 1)", col, err)
	}
	if err := d.SwitchToDefaultContent(); err != nil {
		t.Fatalf("SwitchToDefaultContent: %v", err)
	}
	if _, err := d.FindElement(By.Id("fld")); err != nil {
		t.Fatalf("back in top context, fld should be findable: %v", err)
	}

	// New window / close window.
	before, _ := d.WindowHandles()
	h, err := d.NewWindow("tab")
	if err != nil || h == "" {
		t.Fatalf("NewWindow = %q, %v", h, err)
	}
	if err := d.SwitchToWindow(h); err != nil {
		t.Fatalf("SwitchToWindow: %v", err)
	}
	remaining, err := d.CloseWindow()
	if err != nil {
		t.Fatalf("CloseWindow: %v", err)
	}
	if len(remaining) != len(before) {
		t.Fatalf("after close, %d handles; want %d", len(remaining), len(before))
	}
	d.SwitchToWindow(remaining[0])

	// Submit: walks to the enclosing form and submits (navigates to /framed).
	q, _ := d.FindElement(By.Id("q"))
	if err := q.Submit(); err != nil {
		t.Fatalf("Submit: %v", err)
	}
	if title, _ := d.Title(); title != "Framed" {
		t.Fatalf("after submit title = %q; want Framed", title)
	}

	// PrintPDF -> a base64 PDF (%PDF- header).
	pdf, err := d.PrintPDF(nil)
	if err != nil {
		t.Fatalf("PrintPDF: %v", err)
	}
	raw, err := base64.StdEncoding.DecodeString(pdf)
	if err != nil || len(raw) < 5 || string(raw[:5]) != "%PDF-" {
		t.Fatalf("PrintPDF not a PDF (%d bytes)", len(raw))
	}

	// SetTimeouts unified call.
	if err := d.SetTimeouts(map[string]interface{}{"implicit": 0, "pageLoad": 30000}); err != nil {
		t.Fatalf("SetTimeouts: %v", err)
	}

	t.Logf("full-feature surface: active/exists/css/frame/window/submit/print/timeouts all green")
}

File: ruby/spec/surface_test.rb
# frozen_string_literal: true

# Live surface-coverage test (Ruby): cookies, navigation history, windows, W3C
# actions, screenshot, and execute_script return shapes against a real headless
# Chrome served by a tiny local HTTP server (so cookies/navigation have a real
# http:// origin). Skips if chromedriver is absent. The server is a plain
# TCPServer rather than WEBrick: webrick left Ruby's stdlib in 3.0, and this
# suite must not need a gem to run.

require 'minitest/autorun'
require 'socket'

$LOAD_PATH.unshift File.expand_path('../lib', __dir__)
require 'selenium-webdriver'

class SurfaceTest < Minitest::Test
  PAGE_ONE = '<!doctype html><title>Page One</title>' \
             '<h1 id="hdr">One</h1><a id="go" href="/two">to two</a>' \
             '<button id="btn" onclick="document.getElementById(\'hdr\').textContent=\'clicked\'">b</button>'
  PAGE_TWO = '<!doctype html><title>Page Two</title><h1 id="hdr">Two</h1>'

  def setup
    @driver_bin = which('chromedriver')
    skip 'chromedriver not on PATH' unless @driver_bin

    @web, @web_thread, @web_port = start_content_server

    @port = free_port
    @cd = spawn(@driver_bin, "--port=#{@port}", out: File::NULL, err: File::NULL)
    skip 'chromedriver did not come up' unless wait_up(@port)
  end

  def teardown
    @web&.close
    @web_thread&.kill
    return unless @cd

    Process.kill('TERM', @cd)
    Process.wait(@cd)
  rescue Errno::ESRCH, Errno::ECHILD
    # gone
  end

  def test_surface
    base = "http://127.0.0.1:#{@web_port}"
    d = Selenium::WebDriver.headless_chrome("http://127.0.0.1:#{@port}")
    begin
      d.get("#{base}/one")
      assert_equal 'Page One', d.title

      # navigation history
      d.find_element(Selenium::WebDriver::By::ID, 'go').click
      assert_equal 'Page Two', d.title
      d.back
      assert_equal 'Page One', d.title
      d.forward
      assert_equal 'Page Two', d.title
      d.back

      # cookies
      d.delete_all_cookies
      d.add_cookie('name' => 'flavor', 'value' => 'mint')
      assert_equal 'mint', d.cookie('flavor')['value']
      assert(d.cookies.any? { |c| c['name'] == 'flavor' })
      d.delete_cookie('flavor')
      refute(d.cookies.any? { |c| c['name'] == 'flavor' })

      # windows
      handles = d.window_handles
      assert handles.length >= 1
      assert_includes handles, d.current_window_handle
      d.set_window_rect('width' => 900, 'height' => 650)
      assert_equal 900, d.window_rect['width']

      # execute_script shapes
      assert_equal 42, d.execute_script('return 6*7;')
      assert_equal 'hi', d.execute_script("return 'hi';")
      assert_equal [1, 2, 3], d.execute_script('return [1,2,3];')
      assert_equal({ 'a' => 1 }, d.execute_script('return {a:1};'))
      assert_equal 42, d.execute_script('return arguments[0]+arguments[1];', 40, 2)

      # W3C actions: pointer click on the button.
      btn = d.find_element(Selenium::WebDriver::By::ID, 'btn')
      rect = btn.rect
      cx = (rect['x'] + rect['width'] / 2).to_i
      cy = (rect['y'] + rect['height'] / 2).to_i
      d.perform_actions([{
                          'type' => 'pointer', 'id' => 'mouse',
                          'parameters' => { 'pointerType' => 'mouse' },
                          'actions' => [
                            { 'type' => 'pointerMove', 'duration' => 0, 'x' => cx, 'y' => cy },
                            { 'type' => 'pointerDown', 'button' => 0 },
                            { 'type' => 'pointerUp', 'button' => 0 }
                          ]
                        }])
      assert_equal 'clicked', d.find_element(Selenium::WebDriver::By::ID, 'hdr').text
      d.clear_actions

      # shadow DOM: reach inside an open shadow root, and confirm a non-host
      # element reports no shadow root (W3C code 19).
      d.execute_script(
        "var h=document.createElement('div');h.id='shost';" \
        "document.body.appendChild(h);" \
        "var r=h.attachShadow({mode:'open'});" \
        "r.innerHTML='<p id=\"sinner\">shadowtext</p>';"
      )
      host = d.find_element(Selenium::WebDriver::By::ID, 'shost')
      root = host.shadow_root
      inner = root.find_element(Selenium::WebDriver::By::CSS_SELECTOR, '#sinner')
      assert_equal 'shadowtext', inner.text
      assert_raises(Selenium::WebDriver::NoSuchShadowRootError) do
        d.find_element(Selenium::WebDriver::By::ID, 'hdr').shadow_root
      end

      # screenshot -> PNG
      raw = d.screenshot_base64.unpack1('m')
      assert_equal "\x89PNG".b, raw[0, 4].b
    ensure
      d.quit
    end
  end

  private

  # A two-route static content server on an ephemeral port. Returns
  # [server, thread, port]; the caller closes the server and kills the thread.
  def start_content_server
    server = TCPServer.new('127.0.0.1', 0)
    thread = Thread.new do
      loop do
        begin
          sock = server.accept
        rescue IOError, Errno::EBADF
          break
        end
        Thread.new(sock) do |conn|
          path = conn.gets.to_s.split(' ')[1].to_s
          # Drain the request headers so the client sees a clean response.
          nil while (line = conn.gets) && line.strip != ''
          body = path.start_with?('/two') ? PAGE_TWO : PAGE_ONE
          conn.write("HTTP/1.1 200 OK\r\n" \
                     "Content-Type: text/html; charset=utf-8\r\n" \
                     "Content-Length: #{body.bytesize}\r\n" \
                     "Connection: close\r\n\r\n#{body}")
        rescue StandardError
          nil
        ensure
          begin
            conn.close
          rescue StandardError
            nil
          end
        end
      end
    end
    [server, thread, server.addr[1]]
  end

  def which(cmd)
    ENV['PATH'].split(File::PATH_SEPARATOR).each do |dir|
      p = File.join(dir, cmd)
      return p if File.executable?(p) && !File.directory?(p)
    end
    nil
  end

  def free_port
    s = TCPServer.new('127.0.0.1', 0)
    port = s.addr[1]
    s.close
    port
  end

  def wait_up(port, timeout = 10.0)
    deadline = Time.now + timeout
    while Time.now < deadline
      begin
        TCPSocket.new('127.0.0.1', port).close
        return true
      rescue Errno::ECONNREFUSED
        sleep 0.1
      end
    end
    false
  end
end
